woocommerce/woocommerce-ios
# Order Details: calculate time-consuming order note content `strippedHTML` asynchronously

Refs #789 (will close the issue after verifying this fixes the issue on a device that can reproduce)

## Changes
- Created a generic `AsyncDictionary` as a dictionary wrapper, where the value for a key can be calculated asynchronously, updated on the main thread, and then notifies the callback on update
  - \+ unit test
- In `OrderDetailsDataSource`: after table view sections are calculated, call `orderNote.note.strippedHTML` asynchronously for each order note, and then trigger a table view reload via a new callback property `onReloadTableView: (() -> Void)?` for caller that owns the table view (`OrderDetailsViewController` in this case) to reload UI

## Testing
Prerequisite: have a store with at least one order that has at least one order note
- **NOTE**: unfortunately I wasn't able to reproduce this refresh issue anymore after updating my iPhone XR to iOS 13 (the issue does not occur when built with Xcode 11 beta either). I tried on my iOS 12 iPhone 6s Plus and couldn't reproduce. Instead, I ran time analysis profiling on the simulator and confirmed that the problematic `strippedHTML` is now called in background threads (shown in screenshots below).
- Launch app logged in to the store
- Go to "Orders" tab
- Tap the order row that has at least one order note
- You can scroll down first to verify if the order notes are showing correctly, and then pull down to refresh the order --> the Order Details UI should not be messed up (like labels being misplaced as in #789) - this was reproducible on an iOS 12 iPhone XR device

## Time profiler screenshots

For the same store, same order, after letting the app rest for a bit. Filtered by `strippedHTML` calls

before | this PR
---  | ---
<img width="1396" alt="Screen Shot 2019-07-29 at 10 56 05 AM" src="https://user-images.githubusercontent.com/1945542/62021120-4dfddf80-b1f8-11e9-8db5-bb476480e996.png"> | <img width="1396" alt="Screen Shot 2019-07-29 at 11 03 23 AM" src="https://user-images.githubusercontent.com/1945542/62021121-4e967600-b1f8-11e9-9312-d268ac320988.png">


Update release notes:
